healsparse/io_map_parquet.py

```diff
@@ -0,0 +1,314 @@
+import os
+import numpy as np
+import healpy as hp
+import astropy.io.fits as fits
+
+from .utils import is_integer_value, _compute_bitshift, WIDE_MASK
+from .utils import check_sentinel
+from .healSparseCoverage import HealSparseCoverage
+from .fits_shim import _make_header
+
+use_pyarrow = False
+try:
+    import pyarrow as pa
+    from pyarrow import parquet
+    from pyarrow import dataset
+    use_pyarrow = True
+except ImportError:
+    pass
+
+
+def _read_map_parquet(healsparse_class, filepath, pixels=None, header=False,
+                      degrade_nside=None, weightfile=None, reduction='mean'):
+    """
+    Internal function to read in a HealSparseMap from a parquet dataset.
+
+    Parameters
+    ----------
+    healsparse_class : `type`
+        Type value of the HealSparseMap class.
+    filepath : `str`
+        Name of the file path to read.  Must be a parquet dataset.
+    pixels : `list`, optional
+        List of coverage map pixels to read.
+    header : `bool`, optional
+        Return the parquet metadata as well as map?  Default is False.
+    degrade_nside : `int`, optional
+        Degrade map to this nside on read.  None means leave as-is.
+        Not yet implemented for parquet.
+    weightfile : `str`, optional
+        Floating-point map to supply weights for degrade wmean.  Must
+        be a HealSparseMap (weighted degrade not supported for
+        healpix degrade-on-read).
+        Not yet implemented for parquet.
+    reduction : `str`, optional
+        Reduction method with degrade-on-read.
+        (mean, median, std, max, min, and, or, sum, prod, wmean).
+        Not yet implemented for parquet.
+
+    Returns
+    -------
+    healSparseMap : `HealSparseMap`
+        HealSparseMap from file, covered by pixels
+    header : `astropy.io.fits.Header` (if header=True)
+        Header metadata for the map file.
+    """
+    ds = dataset.dataset(filepath, format='parquet', partitioning='hive')
+    schema = ds.schema
+    # Convert from byte strings
+    md = {key.decode(): schema.metadata[key].decode()
+          for key in schema.metadata}
+
+    if 'healsparse::filetype' not in md:
+        raise RuntimeError("Filepath %s is not a healsparse parquet map." % (filepath))
+    if md['healsparse::filetype'] != 'healsparse':
+        raise RuntimeError("Filepath %s is not a healsparse parquet map." % (filepath))
+    cov_fname = os.path.join(filepath, '_coverage.parquet')
+    if not os.path.isfile(cov_fname):
+        # Note that this could be reconstructed from the information in the file
+        # inefficiently.  This feature could be added in the future.
+        raise RuntimeError("Filepath %s is missing coverage map %s" % (filepath, cov_fname))
+
+    nside_sparse = int(md['healsparse::nside_sparse'])
+    nside_coverage = int(md['healsparse::nside_coverage'])
+    nside_io = int(md['healsparse::nside_io'])
+    bitshift_io = _compute_bitshift(nside_io, nside_coverage)
```

Comment:
Not sure if we need to guard against `nside_io < nside_cov` here.

Reply:
I don't think we can guard against all forms of corrupted files, because such a thing could not have been written in the first place.